src/main/kotlin/ProofSearcher.kt

```diff
@@ -1,22 +1,46 @@
+import analytics.ProofAnalytic
 import parsing.System
 import proofs.Proof
 import proofs.RefinementTransitivity
 import proofs.SelfRefinement
 import proofs.Theorem6Conj2
 
 class ProofSearcher {
-    private val theorems: Array<Proof> = arrayOf(
-        RefinementTransitivity(),
-        SelfRefinement(),
-        Theorem6Conj2()
-    )
+    private val theorems = ArrayList<Proof>()
+    private val analytics = ArrayList<ProofAnalytic>()
+
+    fun addProof(proof: Proof): ProofSearcher{
+        if (!theorems.contains(proof)){
+            theorems.add(proof)
+        }
+
+        return this
+    }
+
+    fun addAnalytic(analytic: ProofAnalytic): ProofSearcher{
+        analytics.add(analytic)
+        return this
+    }
 
     fun findNewRelations(components: ArrayList<System>): ArrayList<System> {
+
+        analytics.forEach{analytic -> analytic.recordStart(components)}
+
+        val result = runSearch(components)
+
+        analytics.forEach{analytic -> analytic.recordEnd(components)}
+
+        return result
+    }
+
+    private fun runSearch(components: ArrayList<System>): ArrayList<System> {
         val allComponents = ArrayList<System>(components)
         var dirtyComponents = HashSet(components)
 
         while (dirtyComponents.isNotEmpty()) {
             dirtyComponents = searchIteration(dirtyComponents, allComponents)
+
+            analytics.forEach{analytic -> analytic.recordIteration(dirtyComponents, components)}
```

Comment:
should it not be `analytic.recordIteration(dirtyComponents, allComponents)` as they may change and we may want to record that?